rust/kcl-lib/src/execution/exec_ast.rs

```diff
@@ -918,165 +918,193 @@ impl Node<MemberExpression> {
 impl Node<BinaryExpression> {
     #[async_recursion]
     pub async fn get_result(&self, exec_state: &mut ExecState, ctx: &ExecutorContext) -> Result<KclValue, KclError> {
-        let left_value = self.left.get_result(exec_state, ctx).await?;
-        let right_value = self.right.get_result(exec_state, ctx).await?;
-        let mut meta = left_value.metadata();
-        meta.extend(right_value.metadata());
-
-        // First check if we are doing string concatenation.
-        if self.operator == BinaryOperator::Add {
-            if let (KclValue::String { value: left, meta: _ }, KclValue::String { value: right, meta: _ }) =
-                (&left_value, &right_value)
-            {
-                return Ok(KclValue::String {
-                    value: format!("{}{}", left, right),
-                    meta,
-                });
-            }
+        let mut partial = self.right.get_result(exec_state, ctx).await?;
+        let mut next = &self.left;
+        let source_range = self.into();
+        // Don't recurse through a big chain of binary operations, iterate instead.
+        while let BinaryPart::BinaryExpression(next_binary_expr) = next {
+            let metadata = partial.metadata();
+            partial = do_binary_op(
+                partial,
+                next_binary_expr.right.get_result(exec_state, ctx).await?,
```

Comment:
Here and above we're still recursing down the right hand side.

I'm not sure about precedence, that shouldn't be an issue because it should be unambiguous after parsing, but I'm not exactly clear about the order of operations here. Because KCL is mostly immutable, I don't think evaluation order is important, though I would rather it were preserved with this refactoring just in case there is an interesting edge case with tags or something.

I would expect to implement this with basically a manual stack - using a Vec rather than the Rust stack for the KCL call stack. Then it's just an iterative depth-first traversal of the AST and we maintain evaluation order